src/node_api.cc

```diff
@@ -3570,3 +3574,435 @@ napi_status napi_run_script(napi_env env,
+  if (status != napi_ok) {
+    return status;
+  }
+
+  status = napi_function_expected;
+
+  if (func_type == napi_function) {
+    ts_fn = new napi_threadsafe_function__;
+    if (ts_fn == nullptr) {
+      status = napi_generic_failure;
+    } else {
+      status = napi_add_env_cleanup_hook(env, napi_cleanup_threadsafe_function,
+          ts_fn);
+      if (status == napi_ok) {
+        if (uv_async_init(env->loop, &ts_fn->async,
+            napi_threadsafe_function_cb) == 0) {
+          ts_fn->ref = v8impl::Reference::New(env,
+              v8impl::V8LocalValueFromJsValue(func).As<v8::Object>(),
+              1, false, nullptr, nullptr, nullptr);
+          if (ts_fn->ref != nullptr) {
+            if (max_queue_size > 0) {
+              ts_fn->cond.reset(new node::ConditionVariable);
+            }
+            if (max_queue_size == 0 || ts_fn->cond) {
+              if (uv_idle_init(env->loop, &ts_fn->idle) == 0) {
+                napi_status async_creation_status =
+                    napi_async_init(env, async_resource, async_resource_name,
+                        &(ts_fn->async_context));
+                if (async_creation_status == napi_ok) {
+                  ts_fn->call_js_cb = call_js_cb == nullptr ?
+                      napi_threadsafe_function_default_call_js : call_js_cb;
+                  ts_fn->context = context;
+                  ts_fn->env = env;
+                  ts_fn->max_queue_size = max_queue_size;
+                  ts_fn->thread_count = initial_thread_count;
+                  ts_fn->is_closing = false;
+                  ts_fn->idle_running = false;
+                  ts_fn->thread_finalize_callback = thread_finalize_cb;
+                  ts_fn->thread_finalize_data = thread_finalize_data;
+                  ts_fn->handles_closed = false;
+
+                  *result = reinterpret_cast<napi_threadsafe_function>(ts_fn);
+                  return napi_clear_last_error(env);
+                }
+                status = async_creation_status;
+              }
+            }
+            v8impl::Reference::Delete(ts_fn->ref);
+          }
+          node::Environment::GetCurrent(env->isolate)->CloseHandle(
+              reinterpret_cast<uv_handle_t*>(&ts_fn->async),
+              [] (uv_handle_t* handle) -> void {
+                delete handle;
+              });
+
+          // Prevent the thread-safe function from being deleted here, because
+          // the callback above will delete it.
+          ts_fn = nullptr;
+        }
+      }
+      delete ts_fn;
+    }
+  }
+  return napi_set_last_error(env, status);
+}
+
+NAPI_EXTERN napi_status
+napi_get_threadsafe_function_context(napi_threadsafe_function ts_fn,
+                                     void** result) {
+  CHECK(ts_fn != nullptr);
+  CHECK(result != nullptr);
+
+  *result = ts_fn->context;
+  return napi_ok;
+}
+
+static napi_status
+napi_tick_threadsafe_function(napi_threadsafe_function ts_fn,
```

Comment:
Also something that threw me off a lot: The name here should reflect that the main purpose is modifying the ref count